BM/tools/pcie/pcie_check.c

```diff
@@ -0,0 +1,1225 @@
+		if (!strstr(line, mmio_bar))
+			continue;
+
+		if (sscanf(line, "%p-%s", &start, base_end) != 2)
+			continue;
+
+		printf("start:%p, base_end:%s\n", start, base_end);
+		if (!start) {
+			printf("BAR(start) is NULL, did you use root to execute?\n");
+			exit(1);
+		}
+		printf("BAR(Base Address Register) for mmio MMCONFIG:%p\n", start);
+		BASE_ADDR = (unsigned long)start;
+		break;
+	}
+	fclose(maps);
+
+	if (BASE_ADDR == 0) {
+		//printf("Check kconfig CONFIG_IO_STRICT_DEVMEM or v6.9 or newer kernel!\n");
+		BASE_ADDR = find_base_from_dmesg();
+		if (BASE_ADDR == 0) {
+			BASE_ADDR = find_base_from_mcfg();
+			if (BASE_ADDR == 0) {
+				printf("No MMIO in dmesg, /proc/iomem and mcfg, check acpidump.\n");
+				exit(2);
+			}
+		}
+	}
+#endif
+	return 0;
+}
+
+void typeshow(u8 data)
+{
+	printf("\tpcie type:%02x  - ", data);
+	switch (data) {
+	case 0x00:
+		printf("PCI Express Endpoint device\n");
+		break;
+	case 0x01:
+		printf("Legacy PCI Express Endpoint device\n");
+		break;
+	case 0x04:
+		printf("RootPort of PCI Express Root Complex\n");
+		break;
+	case 0x05:
+		printf("Upstream Port of PCI Express Switch\n");
+		break;
+	case 0x06:
+		printf("Downstream Port of PCI Express Switch\n");
+		break;
+	case 0x07:
+		printf("PCI Express-to-PCI/PCI-x Bridge\n");
+		break;
+	case 0x08:
+		printf("PCI/PCI-xto PCi Express Bridge\n");
+		break;
+	case 0x09:
+		printf("Root Complex Integrated Endpoint Device\n");
+		break;
+	case 0x0a:
+		printf("Root Complex Event Collector\n");
+		break;
+	default:
+		printf("reserved\n");
+		break;
+	}
+}
+
+void speedshow(u8 speed)
+{
+	printf("\tspeed: %x   - ", speed);
+	switch (speed) {
+	case 0x00:
+		printf("2.5GT/S");
+		break;
+	case 0x02:
+		printf("5GT/S");
+		break;
+	case 0x04:
+		printf("8GT/S");
+		break;
+	default:
+		printf("reserved");
+		break;
+	}
+	printf("\n");
+}
+
+void linkspeed(u8 speed)
+{
+	printf("\tlink speed:%x   - ", speed);
+	switch (speed) {
+	case 0x01:
+		printf("SupportedLink Speeds Vector filed bit 0");
+		break;
+	case 0x02:
+		printf("SupportedLink Speeds Vector filed bit 1");
+		break;
+	case 0x03:
+		printf("SupportedLink Speeds Vector filed bit 2");
+		break;
+	case 0x04:
+		printf("SupportedLink Speeds Vector filed bit 3");
+		break;
+	case 0x05:
+		printf("SupportedLink Speeds Vector filed bit 4");
+		break;
+	case 0x06:
+		printf("SupportedLink Speeds Vector filed bit 5");
+		break;
+	case 0x07:
+		printf("SupportedLink Speeds Vector filed bit 6");
+		break;
+	default:
+		printf("reserved");
+		break;
+	}
+	printf("\n");
+}
+
+void linkwidth(u8 width)
+{
+	printf("\tlink width:%02x - ", width);
+	switch (width) {
+	case 0x01:
+		printf("x1");
+		break;
+	case 0x02:
+		printf("x2");
+		break;
+	case 0x04:
+		printf("x4");
+		break;
+	case 0x08:
+		printf("x8");
+		break;
+	case 0x0c:
+		printf("x12");
+		break;
+	case 0x10:
+		printf("x16");
+		break;
+	case 0x20:
+		printf("x32");
+		break;
+	default:
+		printf("reserved");
+		break;
+	}
```

Comment:
If you like following change:
```
if(width > 0 && width < 33 &&  (width & width - 1) == 0)
        printf("x%d", width);
else
        printf("reserved");
```

Reply:
Yes, it's better.
Will update it.

Thanks!